### PR TITLE
RSDK-6530: Add depth map encode decode methods to C++ SDK

### PR DESCRIPTION
**Pull request by hexbabe**
From https://viam.atlassian.net/browse/RSDK-6530 :

> Sean Yu wrote a deserialize_depth_map function in order to consume depth maps in cpp-land for the [rgb-d-overlay](https://github.com/viam-labs/rgb-d-overlay) module. It is best to make this a static Camera method in C++ SDK

### CODE REVIEW

File: src/viam/sdk/components/camera/camera.hpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.hpp:88]: I think we should. It will require writing an equivalent `encode_depth_map` function, but I think that's a good idea I will do it

File: src/viam/sdk/components/camera/camera.cpp
Comment (hexbabe) [inline src/viam/sdk/components/camera/camera.cpp:29]: Does this stay here, or move into the `Camera::depth_map` constructor? Not sure if assertions like this belong in structs

Comment (stuqdog) [reply src/viam/sdk/components/camera/camera.cpp]: That's a good question! My instinct is to keep it here, I think it makes documenting the error cases more consistent with what's happening in the code base currently. But I'm actually not sure what's more idiomatic for C++ broadly speaking, @acmorrow likely knows more!

**Comment by stuqdog** [comment]
> Wait, does the `run-tests` action run unit tests?

yep! it's a little funky though, if one test fails in a file then often it says they all failed. Lemme know if you want any help with debugging.

**Review by purplenicole730 — COMMENTED** [review]
I'm removing the request for my review since although I read through it, the camera logic is lost on me.

File: src/viam/sdk/components/camera/camera.hpp
Comment (stuqdog) [inline src/viam/sdk/components/camera/camera.hpp:125]: `@param` is now just a `depth_map`

File: src/viam/sdk/components/camera/camera.hpp
Comment (stuqdog) [inline src/viam/sdk/components/camera/camera.hpp:9]: Do we actually need these new includes? I don't see where they're being used here.

In particular the `arpa/inet.h` include is something I'd like to avoid if possible, since it isn't part of the standard library and so creates an additional dependency.

Note it's possible I'm being overly paranoid here! I'm not sure how standard the `arpa/inet.h` include is, so happy to have some pushback on this :)

File: src/viam/sdk/components/camera/camera.cpp
Comment (stuqdog) [reply src/viam/sdk/components/camera/camera.cpp:40]: Ahh another reason to do it here is, someone could theoretically modify the `depth_map` after constructing it, thus calling here with an invalid depth_map even if they constructed a valid one. So yeah, let's leave it as is!

File: src/viam/sdk/components/camera/camera.cpp
Comment (stuqdog) [inline src/viam/sdk/components/camera/camera.cpp:16]: Ahh I see where the `arpa` include comes into play here. If we must `include` it I'd prefer it to be in the `cpp` file rather than the `hpp`. But, it looks like we can do `endian` conversions with [boost](https://www.boost.org/doc/libs/1_84_0/libs/endian/doc/html/endian.html#conversion), which we already are relying on. can we swap to that?

**Review by stuqdog — COMMENTED** [review]
Very cool, though I'll admit the camera logic flew past me a bit! Just a small comment update and a request to swap to `boost` for endian conversion, otherwise looks good!

File: src/viam/sdk/components/camera/camera.cpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.cpp:16]: Good call. I didn't realize boost could do that. Might be best to refactor the Realsense to use that too... diff conversation though

File: src/viam/sdk/components/camera/camera.hpp
Comment (stuqdog) [inline src/viam/sdk/components/camera/camera.hpp:7]: can we put this `include` in the `.cpp`? Seems like an implementation detail that users don't need to know about.

Also note that the `include`s are sorted into different groups that might be a bit unintuitive. This should go with the `google/protobuf` includes in the `.cpp` as a third party include.

File: src/viam/sdk/components/camera/camera.hpp
Comment (stuqdog) [inline src/viam/sdk/components/camera/camera.hpp:133]: Hmm... I think probably this should be a non-static method on `depth_map`. While I like the shorter path of the method being on `Camera`, it seems a bit strange to have it be a static method that only takes a single `depth_map` object as an argument. And if `depth_map` ever were used by other resources such that we wanted to decouple it from `camera`, it would make sense to move the `encode`/`decode` methods along with it.

If we do end up going that route, it would make sense for `decode_depth_map` to similarly move onto `depth_map` (though of course that must remain static!).

**Review by stuqdog — APPROVED** [review]
Nice! A couple small fixes requested (no need to re-request review unless you're uncertain on any of it or have pushback), otherwise LGTM!

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.hpp:100]: That's a neat doxygen trick I didn't know.

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.hpp:102]: The `depth_map` type is the vocabulary type for the SDK. So I think it is worth ensuring that the interface that it presents offers good UX. As written, I still need to do index computation in order to obtain the depth value at a given `(x, y)` by doing `dm.depth_values[dm.height * x + y]` or something like that.

The C++ SDK already has a dependency on the xtensor library via the mlmodel service API. You could trivially represent this 2d array of uint16 with xtensor, and then you would get indexed access for free.

If xtensor feels like overkill for a 2d array there are probably other ways we can get a good 2d array concept rather than pushing the need to index into a linear array to the user.

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [reply src/viam/sdk/components/camera/camera.hpp:133]: Also interested to know if this needs to be part of the public API? It feels like an implementation detail. That might argue for keeping it a free function rather than part of `depth_map` (if that class even continues to exist, per my xtensor comment).

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.hpp:145]: Ditto: not sure if this needs to be part of the public API or not.

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:32]: `size--` looks like a C++ expression here. I'd write this as `"Incorrect data size: attempted ..."`

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:28]: Just curious, but what "format" are we following here? How did big-endian get into the picture (hah!) at all?

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:29]: Utility functions like this should be in the unnamed namespace so they don't have linkage.

**Review by acmorrow — CHANGES_REQUESTED** [review]

**Comment by acmorrow** [comment]
Just a general feeling here, but this seems like a bit of a leaky abstraction. For instance, why are the depths uint16_t's? It feels like that's an artifact of an implementation somewhere, but is somewhat unnatural in the UX. Ditto the flat array presented with height/width. Also unclear to me where the endian-ness aspect comes in.

File: src/viam/sdk/components/camera/camera.cpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.cpp:28]: Here is the context that I have: one of the reasons we are concerned with endianness is because the [Intel Realsense returns big-endian by default](https://github.com/viamrobotics/viam-camera-realsense?tab=readme-ov-file#attributes). Similarly, [the Python SDK also assumes big-endian when decoding](https://github.com/hexbabe/sean-viam-python-sdk/blob/5accda204293685954eb7382859f97c7f57b964b/src/viam/media/video.py#L43). I think @bhaney has more accurate and precise info on why we do this.

File: src/viam/sdk/components/camera/camera.cpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.cpp:29]: Thanks for the review! New to C++ so learning a lot :D

**Comment by hexbabe** [comment]
> Just a general feeling here, but this seems like a bit of a leaky abstraction. For instance, why are the depths uint16_t's? It feels like that's an artifact of an implementation somewhere, but is somewhat unnatural in the UX. Ditto the flat array presented with height/width. Also unclear to me where the endian-ness aspect comes in.

https://github.com/viamrobotics/rdk/blob/5df6d046388e79ab499a2f476d7e2b1fcc09d9e8/rimage/depth_map_raw.go#L311

Above is the specification in RDK about our custom depth map format. I can add a similar blurb about it in `camera.cpp`?

File: src/viam/sdk/components/camera/camera.hpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.hpp:102]: I may be wrong but I think we should keep it as a flat vector. ML frameworks like pytorch and tensorflow has utilities to convert vectors to the native data structs e.g. `torch.from_blob` or `tf.convert_to_tensor`. And then you can use something like `.reshape` with the width and height. I looked it up and it's actually quite a bit harder to make this conversion with xtensor. 

I feel like direct indexing is not a common use case with depth maps? Again, not a CV guy could be wrong lol

File: src/viam/sdk/components/camera/camera.hpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.hpp:145]: > not sure if this needs to be part of the public API or not.

Would the alternative to being part of the public API be to only include it in the cpp file as a private function? Would it just be part of get_image i.e. hook it up to the logic at the end to output the decoded value instead of the current raw bytes? That would be a breaking change then

File: src/viam/sdk/components/camera/camera.cpp
Comment (bhaney) [reply src/viam/sdk/components/camera/camera.cpp:28]: endianness in this context is the endian-ness of a 16bit number, not of a system. Each pixel in the depth image is 2 bytes. For the two bytes, you can either read [first second] or [second first] within the byte stream. When I talk about big-endian, I mean it within this context

Comment (bhaney) [reply src/viam/sdk/components/camera/camera.cpp:28]: calling `$xxd -g 1 test_image.viam.dep` on a test depth file, where a 20 x 10 image has each pixel value as x_location*y_location, this is what you would expect as the byte stream (with a space between every byte):
```
00000000: 44 45 50 54 48 4d 41 50 00 00 00 00 00 00 00 14  DEPTHMAP........
00000010: 00 00 00 00 00 00 00 0a 00 00 00 00 00 00 00 00  ................
00000020: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
00000030: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
00000040: 00 00 00 01 00 02 00 03 00 04 00 05 00 06 00 07  ................
00000050: 00 08 00 09 00 0a 00 0b 00 0c 00 0d 00 0e 00 0f  ................
00000060: 00 10 00 11 00 12 00 13 00 00 00 02 00 04 00 06  ................
00000070: 00 08 00 0a 00 0c 00 0e 00 10 00 12 00 14 00 16  ................
00000080: 00 18 00 1a 00 1c 00 1e 00 20 00 22 00 24 00 26  ......... .".$.&
```
`44 45 50 54 48 4d 41 50` is the string DEPTHMAP in ascii
`00 00 00 00 00 00 00 14` says the image has a width of 20 pixels
`00 00 00 00 00 00 00 0a` says the image has a height of 10 pixels

These need to be big-endian encoded numbers too, because if instead you encoded the width as `14 00 00 00 00 00 00 00`, that would imply an image width of 1441151880758558720

**Review by bhaney — DISMISSED** [review]
Thanks for adding this! Just some small changes for clarity

**Comment by acmorrow** [comment]
@hexbabe -

I think I had sort of misunderstood the point of this review. I expected it was creating a new type like `raw_image` that would actually be _used_ in the `Camera` api (e.g. `raw_image Camera::get_image(...)`). But that isn't what is going on here: the `depth_map` type isn't returned or consumed by `Camera`, and the two new functions exist only to transform a byte buffer into a `depth_map` and vice versa.

So, what is the flow of control here? How does the user of the SDK end up with a byte buffer which happens to contain an encoded depth map, and know that it should be passed to `Camera::decode_depth_map`? Or is this just a "part 1" review and subsequent changes will add methods that traffic in `depth_map` to `Camera`?

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:31]: You can get rid of a lot of these explicit `8`'s by using `sizeof`

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [reply src/viam/sdk/components/camera/camera.hpp:88]: These constants belong in the `.cpp` file. Also you don't need to use `SHOUTY` constants in C++. Often in the C++SDK we prefix constants with `k_` to indicate that they are constant.

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [reply src/viam/sdk/components/camera/camera.hpp:133]: I take back my comment about it being part of the public API. My comment was based on my assumption that `Camera` methods were actually going to be trading in `depth_map`s, in which case I wouldn't have expected this part of the transformation pair to need to be visible to users. However, per my newer understanding about what is being done here, I agree that it needs to be public. I still have some questions though about how users are expected to know what is and isn't a buffer containing a depth map that they should be passing to `Camera::decode_depth_map`, or what they are expected to do with a buffer containing an encoded depth map after they call `Camera::encode_depth_map`.

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [reply src/viam/sdk/components/camera/camera.hpp:145]: As above.

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:54]: Couldn't the `depth_map` constructor enforce this invariant?

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:66]: Might be better if `offset` was passed by pointer here to make it clearer to a reader that it was going to be modified.

File: src/viam/sdk/components/camera/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.cpp:115]: I'll bet you can write this entire thing as just `return {width, height, std::move(arr)};`

File: src/viam/sdk/components/camera/camera.hpp
Comment (acmorrow) [inline src/viam/sdk/components/camera/camera.hpp:99]: There are invariants for a valid depth_map, notably that depth_values.size() == width * height. That'd argue for making this a class so that the invariant can be enforced on construction and then preserved.

**Review by acmorrow — CHANGES_REQUESTED** [review]
Overall, the code looks good. I have some small comments and suggestions.

I do wonder whether 1) the depth_map stuff belongs on `Camera` at all given that the rest of the Camera API doesn't make use of it in any way (yet?), and 2) how exactly it is expected for clients or module authors to use these APIs.

**Comment by hexbabe** [comment]
> @hexbabe -
> 
> I think I had sort of misunderstood the point of this review. I expected it was creating a new type like `raw_image` that would actually be _used_ in the `Camera` api (e.g. `raw_image Camera::get_image(...)`). But that isn't what is going on here: the `depth_map` type isn't returned or consumed by `Camera`, and the two new functions exist only to transform a byte buffer into a `depth_map` and vice versa.
> 
> So, what is the flow of control here? How does the user of the SDK end up with a byte buffer which happens to contain an encoded depth map, and know that it should be passed to `Camera::decode_depth_map`? Or is this just a "part 1" review and subsequent changes will add methods that traffic in `depth_map` to `Camera`?

The scope of this PR is to put decode and encode methods into the SDK as sort of util methods for depth map handling in modules and clients. For decoding, I wanted to mirror how they are handled in the Python SDK. In the Python SDK, we leave it to the client to call `raw_img.bytes_to_depth_array()` to decode the depth map. The goal here is to provide similar decode util for C++ SDK users. For encoding, we would expect depth camera module creators to use the encode function to convert their raw bytes into the Viam depth mime type format.

Sorry for all the confusion. My mistake was thinking that this ticket would be as simple as copying over [the function I wrote for the RGB-D overlay module](https://github.com/viam-labs/rgb-d-overlay/blob/4d102284647d39d62bfd47bc44b14dae3b7f6ccd/src/main.cpp#L29) into `Camera` to be used as util functions. If I were to redo this PR, I would've communicated the requirements and thought harder about how to integrate the methods into the exiting C++ APIs.

To amend this mistake, I think we should chat briefly to discuss where it makes sense to place these functions in a less haphazard way.

**Comment by acmorrow** [comment]
@hexbabe and I met to discuss this. The plan is to move forward with this review. The `depth_map` type will be re-worked in terms of `xtensor` (either it will be an `xtensor` type alias, or it will wrap an `xtensor`). There will then be a subsequent review where the Camera interface is adjusted in order to integrate depth map information. This might look like having `Camera::get_image` return a variant over `raw_image` and `depth_map`, or adding a `get_depth_map` function that returns an `Optional<depth_map>`. Similar adjustments will need to be made to `get_images` to incorporate `depth_map`s.

File: src/viam/sdk/components/camera/camera.hpp
Comment (hexbabe) [reply src/viam/sdk/components/camera/camera.hpp:88]: From the gcc [manual](http://gcc.gnu.org/onlinedocs/gcc/Long-Long.html):

> ISO C99 supports data types for integers that are at least 64 bits wide ( . . . ) . To make an integer constant of type long long int, add the suffix LL to the integer. To make an integer constant of type unsigned long long int, add the suffix ULL to the integer.

It should ensure that the literal has the correct type and size @bhaney

File: src/viam/sdk/components/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera.cpp:88]: If you make `append_uint64_big_endian` a template on the value type, you could re-use it here:

```
template<typename T>
append_big_endian(std::vector<unsigned char>& data, size_t* offset, T value);
```

File: src/viam/sdk/components/camera.cpp
Comment (acmorrow) [inline src/viam/sdk/components/camera.cpp:129]: I think this is going to cause problems: I think `xt::adapt` is only going to view the memory in `depth_values`, and it will go out of scope when this function returns.

I think you will need another approach to construct an xtensor that owns the memory here. Let me know if you need me to look into how to do that best.

**Review by acmorrow — CHANGES_REQUESTED** [review]
This looks really good, though I think there is a memory error that needs to be addressed.

File: src/viam/sdk/components/camera.cpp
Comment (hexbabe) [reply src/viam/sdk/components/camera.cpp:88]: Wow that's so elegant it brings a tear to my eye

File: src/viam/sdk/components/camera.cpp
Comment (hexbabe) [inline src/viam/sdk/components/camera.cpp:122]: I looked up a method to copy the memory in. Let me know if there's a better way to do it @acmorrow

Comment (acmorrow) [reply src/viam/sdk/components/camera.cpp:122]: Hey, I was actually just looking at that. So, what you have here is fine, though it does cost you a copy. If you wanted to avoid the copy though, I think that's totally doable.

The trick would be to not create a `std::vector`, but just to directly create the `xt::xarray` you want to return, of the appropriate size and shape, before the loop, and then populate it in the loop.

```
xt::xarray<uint16_t> m = xt::xarray<uint16_t>::from_shape({height, width});
for (size_t i = 0; i < width * height; ++i) {
    m(<ii>, <jj>) = read_big_endian<uint16_t>(data, &offset);
```

Now, that'd require a little hassle to compute the correct `ii` and `jj` w.r.t. `height` and `width` and `i`, which is sort of a hassle. However, `xtensor` lets you reshape and view. So I think you could do it something like this:

```
xt::xarray<uint16_t> m = xt::xarray<uint16_t>::from_shape({height, width});
auto m_linear_view = xt::flatten(m);
for (size_t i = 0; i < width * height; ++i) {
    m_linear_view[i] = read_big_endian<uint16_t>(data, &offset));
return m;
```

It may not be exactly `flatten`, you might need `reshape_view`? I'd need to spend a little time with the `xtensor` docs to be sure. But something from https://xtensor.readthedocs.io/en/latest/view.html.

I'd say spend no more than 15 minutes on it. If you can make it work, great. If not, put in a TODO and leave it as a copy.

**Review by acmorrow — APPROVED** [review]
LGTM if you want to keep the copy. If you elect to use reshape/view to build directly into the output tensor, please post an update with that so I can give it a look.

**Pull request merged**
